### ISSUE TITLE
Horovod.torch does not reproduce not distributed training

### ISSUE THREAD

**Issue by boydad** [horovod/horovod#3743]
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.12.0a0+git664058f
3. Horovod version: 0.25.0
4. MPI version: mpiexec version 1.1.7
5. CUDA version: Driver Version: 470.103.01   CUDA Version: 11.4 
6. NCCL version: 2.10.3
7. Python version: 3.8.13
8. Spark / PySpark version:
9. Ray version:
10. OS and version: SUSE Linux Enterprise Server 15 SP3
11. GCC version:
12. CMake version:

I runned training scripts without distributed training, training with PyTorch DDP with 1 worker, and horovod with 1 worker. Not distributed version perfectly agreed with 1 worker PT DDP while horovod has a different training curve just after the first gradient update. [I've dug in sources](https://github.com/horovod/horovod/blob/4281bd7212136d8c201f90f64924ed515a796383/horovod/torch/functions.py#L95) and found that horovod initialized optimizer parameters and calls `step()` method. This behavior changed the optimizer state, and makes training more stable, but explicitly breaks reproductivity.

**Comment by chongxiaoc** [comment]
Is there a simple reproducer you can provide?
A simple toy model with dummy data for example?
